File: app/src/main/java/com/erkaslan/moviedb/ui/MovieListAdapter.java
package com.erkaslan.moviedb.ui;

import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.core.graphics.drawable.RoundedBitmapDrawable;
import androidx.core.graphics.drawable.RoundedBitmapDrawableFactory;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.target.BitmapImageViewTarget;
import com.erkaslan.moviedb.data.Movie;
import java.util.ArrayList;
import com.erkaslan.moviedb.R;

public class MovieListAdapter extends RecyclerView.Adapter {
    ArrayList<Movie> movieList;
    Context context;

    public MovieListAdapter(Context context, ArrayList<Movie> movieList) {
        this.movieList = movieList;
        this.context = context;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_movie, parent, false);
        return new MovieListHolder(view, context);
    }

    //Creates a movie and puts movie name and sets a click listener
    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        final Movie movie = (Movie) movieList.get(position);
        ((MovieListHolder) holder).bind(movie);

        ((MovieListHolder) holder).cardMovie.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Bundle bundle = new Bundle();
                bundle.putString("title", movie.getTitle());
                bundle.putString("imdbID", movie.getImdbId());
                bundle.putString("poster", movie.getPoster());
                bundle.putString("year", movie.getYear());
                Navigation.findNavController(v).navigate(R.id.action_homeFragment_to_detailFragment, bundle);
            }
        });


    }

    @Override
    public int getItemCount() {
        return movieList.size();
    }

    class MovieListHolder extends RecyclerView.ViewHolder{
        TextView textTitle,textYear;
        ImageView imageMovie;
        CardView cardMovie;
        Context context;

        public MovieListHolder(View view, Context context) {
            super(view);
            this.context = context;
            textTitle = (TextView) view.findViewById(R.id.textTitle);
            textYear = (TextView) view.findViewById(R.id.textYear);
            imageMovie = (ImageView) view.findViewById(R.id.imageMovie);
            cardMovie = (CardView) view.findViewById(R.id.cardMovie);
        }

        void bind(Movie movie) {

            textTitle.setText("Title: " + movie.getTitle());
            textYear.setText("Year: " + movie.getYear());
            Uri uri = Uri.parse(movie.getPoster());
            displayImageFromUrl(context, uri, imageMovie);

        }

    }

    //Loads and displays an image in circular shape
    public static void displayImageFromUrl(final Context context, final Uri uri, final ImageView imageView) {
        Glide.with(context)
                .asBitmap()
                .load(uri)
                .override(500)
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                //.centerCrop()
                .dontAnimate()
                .placeholder(R.drawable.ic_nomovie_background)
                .into(imageView);/* Round image option
                .into(new BitmapImageViewTarget(imageView) {
                    @Override
                    protected void setResource(Bitmap resource) {
                        RoundedBitmapDrawable circularBitmapDrawable =
                                RoundedBitmapDrawableFactory.create(context.getResources(), resource);
                        circularBitmapDrawable.setCircular(true);
                        imageView.setImageDrawable(circularBitmapDrawable);
                    }
                })*/
    }

}

File: app/src/main/java/com/erkaslan/moviedb/ui/MovieListAdapter.kt
package com.erkaslan.moviedb.ui

import android.content.Context
import android.net.Uri
import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import android.view.LayoutInflater
import com.erkaslan.moviedb.R
import android.util.Log
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.cardview.widget.CardView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.erkaslan.moviedb.data.Movie
import java.util.ArrayList

class MovieListAdapter(var context: Context, var movieList: ArrayList<Movie>, var itemClickListener: ItemClickListener) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    interface ItemClickListener {
        fun itemClick(position: Int, movie: Movie)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        Log.v("ADAPTER", "Home: onCreateViewHolder")
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_movie, parent, false)
        return MovieListHolder(view, context)
    }

    //Creates a movie and puts movie name and sets a click listener
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val movie = movieList[position]
        (holder as MovieListHolder).bind(movie)

        holder.cardMovie.setOnClickListener { v ->
            itemClickListener.itemClick(position, movie)
        }
    }

    override fun getItemCount(): Int {
        return movieList.size
    }

    internal inner class MovieListHolder(view: View, var context: Context) : RecyclerView.ViewHolder(view) {
        var textTitle: TextView
        var textYear: TextView
        var imageMovie: ImageView
        var cardMovie: CardView
        fun bind(movie: Movie) {
            textTitle.text = "Title: " + movie.title
            textYear.text = "Year: " + movie.year
            val uri = Uri.parse(movie.poster)
            displayImageFromUrl(context, uri, imageMovie)
        }

        init {
            textTitle = view.findViewById<View>(R.id.textTitle) as TextView
            textYear = view.findViewById<View>(R.id.textYear) as TextView
            imageMovie = view.findViewById<View>(R.id.imageMovie) as ImageView
            cardMovie = view.findViewById<View>(R.id.cardMovie) as CardView
        }
    }

    companion object {
        //Loads and displays an image in rectangle shape
        fun displayImageFromUrl(context: Context?, uri: Uri?, imageView: ImageView?) {
            Glide.with(context!!)
                .asBitmap()
                .load(uri)
                .override(500)
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .dontAnimate()
                .placeholder(R.drawable.ic_nomovie_background)
                .into(imageView!!)
        }
    }
}